lib/ukvmem/vma_file.c

```diff
@@ -0,0 +1,234 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <stddef.h>
+#include <errno.h>
+
+#include "vmem.h"
+
+#include <uk/config.h>
+#include <uk/assert.h>
+#include <uk/falloc.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/paging.h>
+#ifdef CONFIG_HAVE_PAGING
+#include <uk/plat/paging.h>
+#endif /* CONFIG_HAVE_PAGING */
+#include <vfscore/file.h>
+#include <vfscore/vnode.h>
+#include <vfscore/uio.h>
+#include <uk/isr/string.h>
+
+#ifdef CONFIG_LIBUKVMEM_FILE_BASE
+static __vaddr_t vma_op_file_get_base(struct uk_vas *vas __unused,
+				      void *data __unused,
+				      unsigned long flags __unused)
+{
+	return CONFIG_LIBUKVMEM_FILE_BASE;
+}
+#endif /* CONFIG_LIBUKVMEM_FILE_BASE */
+
+int vma_op_file_new(struct uk_vas *vas, __vaddr_t vaddr __unused,
+		    __sz len __unused, void *data, unsigned long *flags,
+		    struct uk_vma **vma)
+{
+	struct uk_vma_file_args *args = (struct uk_vma_file_args *)data;
+	struct uk_vma_file *vma_file;
+
+	UK_ASSERT(data);
+	UK_ASSERT(args->fd >= 0);
+	UK_ASSERT(args->offset >= 0);
+	UK_ASSERT(PAGE_ALIGNED(args->offset));
+
+	/* Shared mappings are not supported */
+	if (*flags & UK_VMA_FILE_SHARED)
+		return -ENOTSUP;
+
+	/* Since we cannot do ISR-safe file accesses in the fault handler,
+	 * we enforce full load at mapping time for now.
+	 *
+	 * TODO: Remove this restriction if possible.
+	 */
+	*flags |= UK_VMA_MAP_POPULATE;
+
+	vma_file = uk_malloc(vas->a, sizeof(struct uk_vma_file));
+	if (unlikely(!vma_file))
+		return -ENOMEM;
+
+	vma_file->f = vfscore_get_file(args->fd);
+	if (unlikely(!vma_file->f)) {
+		uk_free(vas->a, vma_file);
+		return -EBADF;
+	}
+	vma_file->offset = args->offset;
+
+	/* Use the file name as VMA name. Since the memory management of the
+	 * string is tied to the file object, we do not need to care about
+	 * freeing it. So it is ok, if the caller should override the name.
+	 */
+	vma_file->base.name = vma_file->f->f_dentry->d_path;
+
+	UK_ASSERT(vma);
+	*vma = &vma_file->base;
+
+	return 0;
+}
+
+static void vma_op_file_destroy(struct uk_vma *vma)
+{
+	struct uk_vma_file *vma_file = (struct uk_vma_file *)vma;
+
+	UK_ASSERT(vma_file->f);
+	fdrop(vma_file->f);
+}
+
+static int vma_file_read(struct vfscore_file *fp, __vaddr_t buf, __sz len,
+			 __off offset, __sz *bytes)
+{
+	struct vnode *vp = fp->f_dentry->d_vnode;
+	struct iovec iovec = {
+		.iov_base = (void *)buf,
+		.iov_len = len,
+	};
+	struct uio uio = {
+		.uio_iov = &iovec,
+		.uio_iovcnt = 1,
+		.uio_offset = offset,
+		.uio_resid = len,
+		.uio_rw = UIO_READ,
+	};
+	int rc;
+
+	vn_lock(vp); {
```

Comment:
This looks like someone wants to use RAII but has to use C. I would say the style does not match the one of Unikraft.

Reply:
You got me 😆 Removed.